[BottomNavigationBar] Add autolayout support

romoore opened the pull request:
Adds support to BottomNavigationBar to more easily work with Autolayout.  

* `intrinsicContentSize` returns the correct size of the bar items as content.
* `barItemsBottomAnchor` allows anchoring the Bottom Navigation bar to the safe area layout guide.

Updates the `MDCBottomNavigationBarController` to use the new layout anchor and eliminates the need to respond to safe area changes in the view controller.

Fixes #4520

Commit:
[BottomNavigationBar] Add autolayout support

material-automation commented on components/BottomNavigation/src/MDCBottomNavigationBar.m:
```diff
@@ -159,18 +157,40 @@ - (void)layoutSubviews {
   }
   [self layoutItemViews];
+  [super layoutSubviews];
+
```
clang-format suggested the following change:


```
-
```

<!-- identifier: clang-format -->
<!-- File: components/BottomNavigation/src/MDCBottomNavigationBar.m -->

material-automation commented on components/BottomNavigation/src/MDCBottomNavigationBar.m:
```diff
@@ -159,18 +157,40 @@ - (void)layoutSubviews {
+
+- (void)safeAreaInsetsDidChange {
+  BOOL needsLayout = !UIEdgeInsetsEqualToEdgeInsets(self.formerSafeAreaInsets,
+                                                    self.mdc_safeAreaInsets);
```
clang-format suggested the following change:


```
-  BOOL needsLayout = !UIEdgeInsetsEqualToEdgeInsets(self.formerSafeAreaInsets,
-                                                    self.mdc_safeAreaInsets);
+  BOOL needsLayout =
+      !UIEdgeInsetsEqualToEdgeInsets(self.formerSafeAreaInsets, self.mdc_safeAreaInsets);
```

<!-- identifier: clang-format -->
<!-- File: components/BottomNavigation/src/MDCBottomNavigationBar.m -->

material-automation commented:
clang-format recommended changes.

Consider installing [git-clang-format](https://github.com/llvm-mirror/clang/blob/master/tools/clang-format/git-clang-format) and running the following command:

```bash
git clang-format $(git merge-base origin/develop HEAD)
```

<!-- identifier: clang-format -->

material-automation commented:
bazel detected changes to the following targets:

```
//components/BottomNavigation:BottomNavigation
//components/BottomNavigation:BottomNavigationBeta
//components/BottomNavigation:ColorThemer
//components/BottomNavigation:ObjcExamples
//components/BottomNavigation:SwiftExamples
//components/BottomNavigation:TypographyThemer
//components/BottomNavigation:snapshot_test_lib
//components/BottomNavigation:snapshot_tests
//components/BottomNavigation:snapshot_tests.swift_runtime_linkopts
//components/BottomNavigation:snapshot_tests_test_binary
//components/BottomNavigation:snapshot_tests_test_bundle
//components/BottomNavigation:unit_test_beta_sources
//components/BottomNavigation:unit_test_sources
//components/BottomNavigation:unit_tests
//components/BottomNavigation:unit_tests_IPAD_PRO_12_9_IN_9_3
//components/BottomNavigation:unit_tests_IPAD_PRO_12_9_IN_9_3.swift_runtime_linkopts
//components/BottomNavigation:unit_tests_IPAD_PRO_12_9_IN_9_3_test_binary
//components/BottomNavigation:unit_tests_IPAD_PRO_12_9_IN_9_3_test_bundle
//components/BottomNavigation:unit_tests_IPHONE_5_IN_8_1
//components/BottomNavigation:unit_tests_IPHONE_5_IN_8_1.swift_runtime_linkopts
//components/BottomNavigation:unit_tests_IPHONE_5_IN_8_1_test_binary
//components/BottomNavigation:unit_tests_IPHONE_5_IN_8_1_test_bundle
//components/BottomNavigation:unit_tests_IPHONE_7_PLUS_IN_10_3
//components/BottomNavigation:unit_tests_IPHONE_7_PLUS_IN_10_3.swift_runtime_linkopts
//components/BottomNavigation:unit_tests_IPHONE_7_PLUS_IN_10_3_test_binary
//components/BottomNavigation:unit_tests_IPHONE_7_PLUS_IN_10_3_test_bundle
//components/BottomNavigation:unit_tests_IPHONE_X_IN_11_0
//components/BottomNavigation:unit_tests_IPHONE_X_IN_11_0.swift_runtime_linkopts
//components/BottomNavigation:unit_tests_IPHONE_X_IN_11_0_test_binary
//components/BottomNavigation:unit_tests_IPHONE_X_IN_11_0_test_bundle
//components/BottomNavigation:unit_tests_beta
//components/BottomNavigation:unit_tests_beta.swift_runtime_linkopts
//components/BottomNavigation:unit_tests_beta_test_binary
//components/BottomNavigation:unit_tests_beta_test_bundle
```

<!-- identifier: affected-targets -->

Commit:
Merge branch 'develop' into issue-4520-b

Commit:
Clean up code.

material-automation commented on components/BottomNavigation/src/MDCBottomNavigationBar.m:
```diff
@@ -159,18 +155,29 @@ - (void)layoutSubviews {
+  CGFloat height = self.isTitleBelowIcon ? kBarHeightStackedTitle : kBarHeightAdjacentTitle;
+  CGFloat itemWidth = [self widthForItemsWhenCenteredWithAvailableWidth:CGFLOAT_MAX height:height];
+  CGSize size =
+      CGSizeMake(itemWidth * self.items.count, height);
```
clang-format suggested the following change:


```
-  CGSize size =
-      CGSizeMake(itemWidth * self.items.count, height);
+  CGSize size = CGSizeMake(itemWidth * self.items.count, height);
```

<!-- identifier: clang-format -->
<!-- File: components/BottomNavigation/src/MDCBottomNavigationBar.m -->

material-automation commented on components/BottomNavigation/src/MDCBottomNavigationBarController.m:
```diff
@@ -223,12 +212,12 @@ - (void)loadConstraints {
   [self.navigationBar.bottomAnchor constraintEqualToAnchor:self.view.bottomAnchor].active = YES;
+  self.bnTopCons = [self.navigationBar.topAnchor
+      constraintEqualToAnchor:self.view.safeAreaLayoutGuide.bottomAnchor
+                    constant:-self.navigationBar.intrinsicContentSize.height];
```
clang-format suggested the following change:


```
-                    constant:-self.navigationBar.intrinsicContentSize.height];
+                     constant:-self.navigationBar.intrinsicContentSize.height];
```

<!-- identifier: clang-format -->
<!-- File: components/BottomNavigation/src/MDCBottomNavigationBarController.m -->

Commit:
Formatting

Commit:
Don't invalidate size so often.

Commit:
Naming/formatting

Commit:
Move `viewSafeAreaInsetsDidChange`

Commit:
Remove extra setNeedsLayout

material-automation commented on components/BottomNavigation/src/MDCBottomNavigationBarController.m:
```diff
@@ -49,14 +47,11 @@ - (void)viewDidLoad {
 - (void)viewSafeAreaInsetsDidChange {
   [super viewSafeAreaInsetsDidChange];
-  [self updateNavigationBarHeight];
+  
```
clang-format suggested the following change:


```
-  
+
```

<!-- identifier: clang-format -->
<!-- File: components/BottomNavigation/src/MDCBottomNavigationBarController.m -->

material-automation commented:
clang-format recommended changes.

Consider installing [git-clang-format](https://github.com/llvm-mirror/clang/blob/master/tools/clang-format/git-clang-format) and running the following command:

```bash
git clang-format $(git merge-base origin/develop HEAD)
```

<!-- identifier: clang-format -->

Commit:
A bit more clean up

Commit:
Restore comment

Commit:
Add unit test

Commit:
Merge branch 'develop' into issue-4520-b

Commit:
Integrate with iOS 8 constraints

Commit:
Merge branch 'develop' into issue-4520-b

Commit:
Fix constant names

Commit:
Add UILayoutGuide for barItems.

Adds a UILayoutGuide specifically for positioning the
BottomNavigationBar's content at the top of the safe area layout guide.
This is a much simpler implementation for clients, who should no longer
need to respond directly to safeAreaInsetsDidChange to recompute
constraint constants.

Commit:
Merge branch 'develop' into issue-4520-b

material-automation commented:
## API diff detected the following changes

### BottomNavigation

#### MDCBottomNavigationBar

*new* property: `barItemsBottomAnchor` in `MDCBottomNavigationBar`

<!-- identifier: apidiff -->

Commit:
Add a bit more detail to the new layout guide's docs

Commit:
Nullability

codeman7 commented on components/BottomNavigation/src/MDCBottomNavigationBar.h:
```diff
@@ -162,6 +162,15 @@ typedef NS_ENUM(NSInteger, MDCBottomNavigationBarAlignment) {
+/**
+ UILayoutGuide encompassing the area around all visible items in the bar.
+
+ Note: It is recommended that the bottom anchor of this layout guide be anchored to the bottom of
```
_Tip:_ `@note`

romoore replied:
Done, thanks!

Commit:
Docs

Commit:
Explain how autoresizing mask is working with layout.

Commit:
Only expose an anchor, not the entire layout guide.

Commit:
Oops! Turn off snapshot recording.
